pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go

```diff
@@ -271,12 +271,20 @@ func (r *Reconciler) runReconcileShootFlow(ctx context.Context, o *operation.Ope
 			Fn:           flow.TaskFn(botanist.DeployEtcd).RetryUntilTimeout(defaultInterval, helper.GetEtcdDeployTimeout(o.Shoot, defaultTimeout)),
 			Dependencies: flow.NewTaskIDs(initializeSecretsManagement, deployCloudProviderSecret, waitUntilBackupEntryInGardenReconciled, waitUntilEtcdBackupsCopied),
 		})
-		_ = g.Add(flow.Task{
+		destroySourceBackupEntry = g.Add(flow.Task{
 			Name:         "Destroying source backup entry",
 			Fn:           botanist.DestroySourceBackupEntry,
 			SkipIf:       !allowBackup,
 			Dependencies: flow.NewTaskIDs(deployETCD),
 		})
+		_ = g.Add(flow.Task{
+			Name: "Waiting until source backup entry has been deleted",
+			Fn: func(ctx context.Context) error {
+				return botanist.Shoot.Components.SourceBackupEntry.WaitCleanup(ctx)
+			},
+			SkipIf:       !allowBackup,
```

Comment:
```suggestion
			SkipIf:       !allowBackup || skipReadiness || !botanist.IsRestorePhase(),
```

Note that the check for whether backup is enabled and whether the reconciliation is currently in `restore` phase for the `Destroying source backup entry` task is done here: https://github.com/gardener/gardener/blob/1c53b439dcc2c8902298f9d7a619c517bf76de69/pkg/gardenlet/operation/botanist/backupentry.go#L98-L100

IIRC this was done back in the day when `botanist.IsRestorePhase()` wasn't an exposed function. 
We could potentially also move those to the `SkipIf` check for consistency.